### ISSUE TITLE
use barlowtwins predict but blocked and can not be killed

### ISSUE THREAD

**Issue by showkeyjar** [facebookresearch/barlowtwins#25]
Hi, I'm very glad to thank your guys share this amazing project.
it training very easy.
but when I use it to test or predict, it blocked and can not be killed

here is my predictor code:

`
import torch
import argparse
from PIL import Image
from torch import nn, optim
from main import BarlowTwins, Transform

"""
BarlowTwins Predictor
"""

def load_model(args):
    args.ngpus_per_node = torch.cuda.device_count()
    args.rank = 0
    args.dist_url = 'tcp://localhost:58472'
    args.world_size = args.ngpus_per_node
    gpu = 0
    model = BarlowTwins(args).cuda(gpu)
    model = nn.SyncBatchNorm.convert_sync_batchnorm(model)
    torch.distributed.init_process_group(
        backend='nccl', init_method=args.dist_url,
        world_size=args.world_size, rank=args.rank)
    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[gpu])
    best_cp = torch.load("model/barlowtwins/bt_face.pth")
    # model.module.backbone.state_dict(best_cp)
    model.load_state_dict(best_cp)
    model.eval()
    return model

def predict(t1, t2, model, device):
    model.to(device)
    with torch.no_grad():
        t1=t1.to(torch.device("cuda:1" if torch.cuda.is_available() else "cpu"))
        t2=t2.to(torch.device("cuda:2" if torch.cuda.is_available() else "cpu"))
        out_data = model(t1, t2)
        return out_data

def predict_img(img_url, args):
    img = Image.open(img_url)
    bt_trans = Transform()
    t1, t2 = bt_trans(img)
    t1 = t1.unsqueeze(0)
    t2 = t2.unsqueeze(0)
    model = load_model(args)
    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
    output = predict(t1, t2, model, device)
    return output

if __name__=="__main__":
    parser = argparse.ArgumentParser(description='Barlow Twins Predict')
    parser.add_argument('--workers', default=8, type=int, metavar='N',
                        help='number of data loader workers')
    parser.add_argument('--projector', default='8192-8192-8192', type=str,
                        metavar='MLP', help='projector MLP')
    parser.add_argument("-d", default="data/test/1.jpg")
    args = parser.parse_args()
    img_url = str(args.d)
    result = predict_img(img_url, args)
    print(result)
`

would you please tell me which part is wrong? I'll very hopefully for your answer!

**Comment by jzbontar** [comment]
Hello @showkeyjar,

Can you tell us a bit more about what you are trying to achieve? Usually, when we think about predicting or testing Barlow Twins, we only use the weights of the resnet50, like in [evaluate.py](https://github.com/facebookresearch/barlowtwins/blob/main/evaluate.py#L81-L83). But in your code snippet, I see that you are loading the full Barlow Twins model.

> but when I use it to test or predict, it blocked and can not be killed

Your code probably blocks in the call to `torch.distributed.init_process_group`. You might need to call `torch.multiprocessing.spawn` (see [main.py](https://github.com/facebookresearch/barlowtwins/blob/main/main.py#L70)) or simply avoid using torch.distributed and use a single gpu, which should be enough for testing. 

You can kill the process by pressing CTRL-\ in your terminal window.

**Comment by showkeyjar** [comment]
@jzbontar  thanks for your reply, I'll try it.

**Comment by showkeyjar** [comment]
I have use resnet50 to load weights, and predict a image result, but it returns a tensor including a float list.
I read the paper, and think that maybe the hidden features's score,
so I want to know how to transfer it to a total loss score or probability score?
the purpose is to use it predict image class, repeat paper results.
Thanks again for your attention!

**Comment by jzbontar** [comment]
If you want to use barlow twins to predict the image class you can do the following:

1. Load the resnet weights that were trained with barlow twins. You can use the pretrained model we released.
2. Randomly initialize the last fully connected layer of the resnet, because it is not trained during barlow twins training.
3. Use a labeled dataset to train the last fully connected layer (and, if you want, finetune the other weights of the resnet)

This is exactly what we do in [evaluate.py](https://github.com/facebookresearch/barlowtwins/blob/main/evaluate.py) so you can use that script as a starting point.
